### PR TITLE
Always quote string

### LINKED ISSUE #125: Failed to read Fannie Mae data
I can read in the [Fannie Mae performance data](https://loanperformancedata.fanniemae.com/lppub/index.html). Unfortunatley, I can't provide the data as you need to register for it.

```julia
julia> using TextParse; @time a = csvread("c:/data/Performance_2016Q1.csv", '|', header_exists=false, type_detect_rows = 200_000)
Error while guessing a common type for column 2
new value: 02/01/2016, prev guess was: TextParse.DateTimeToken{Dates.Date,Dates.DateFormat{Symbol("mm/dd/yyyy"),Tuple{Dates.DatePart{'m'},Dates.Delim{Char,1},Dates.DatePart{'d'},Dates.Delim{Char,1},Dates.DatePart{'y'}}}}(dateformat"mm/dd/yyyy")
prev value: 103764760078
ERROR: InterruptException:
Stacktrace:
 [1] guesscolparsers(::TextParse.VectorBackedUTF8String, ::Array{String,1}, ::TextParse.LocalOpts{UInt8,UInt8,UInt8}, ::Int64, ::Int64, ::Array{Any,1}, ::Type, ::Nothing, ::Array{String,1}, ::Nothing) at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\csv.jl:519
 [2] #_csvread_internal#26(::Bool, ::Char, ::Char, ::Nothing, ::Type, ::Type, ::Bool, ::Int64, ::Nothing, ::Nothing, ::Int64, ::Nothing, ::Bool, ::Array{String,1}, ::Array{String,1}, ::OrderedCollections.OrderedDict{Union{Int64, String},AbstractArray{T,1} where T}, ::Int64, ::Nothing, ::Array{Any,1}, ::String, ::Int64, ::typeof(TextParse._csvread_internal), ::TextParse.VectorBackedUTF8String, ::Char) at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\csv.jl:244
 [3] (::getfield(TextParse, Symbol("#kw##_csvread_internal")))(::NamedTuple{(:filename, :header_exists, :type_detect_rows),Tuple{String,Bool,Int64}}, ::typeof(TextParse._csvread_internal), ::TextParse.VectorBackedUTF8String, ::Char) at .\none:0
 [4] (::getfield(TextParse, Symbol("##22#24")){Base.Iterators.Pairs{Symbol,Integer,Tuple{Symbol,Symbol},NamedTuple{(:header_exists, :type_detect_rows),Tuple{Bool,Int64}}},String,Char})(::IOStream) at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\csv.jl:110
 [5] #open#310(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::getfield(TextParse, Symbol("##22#24")){Base.Iterators.Pairs{Symbol,Integer,Tuple{Symbol,Symbol},NamedTuple{(:header_exists, :type_detect_rows),Tuple{Bool,Int64}}},String,Char}, ::String, ::Vararg{String,N} where N) at .\iostream.jl:369
 [6] open at .\iostream.jl:367 [inlined]
 [7] #_csvread_f#20 at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\csv.jl:107 [inlined]
 [8] #_csvread_f at .\none:0 [inlined]
 [9] #csvread#16(::Base.Iterators.Pairs{Symbol,Integer,Tuple{Symbol,Symbol},NamedTuple{(:header_exists, :type_detect_rows),Tuple{Bool,Int64}}}, ::Function, ::String, ::Char) at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\csv.jl:78
 [10] (::getfield(TextParse, Symbol("#kw##csvread")))(::NamedTuple{(:header_exists, :type_detect_rows),Tuple{Bool,Int64}}, ::typeof(csvread), ::String, ::Char) at .\none:0
 [11] top-level scope at util.jl:156

(v1.1) pkg> status
    Status `C:\Users\RTX2080\.julia\environments\v1.1\Project.toml`
  [c52e3926] Atom v0.7.15
  [336ed68f] CSV v0.4.3
  [5d742f6a] CSVFiles v0.14.0
  [a93c6f00] DataFrames v0.17.1
  [1313f7d8] DataFramesMeta v0.4.1
  [5789e2e9] FileIO v1.0.5
  [c91e804a] Gadfly v1.0.1
  [cd3eb016] HTTP v0.7.1
  [7073ff75] IJulia v1.17.0
  [e5e0dc1b] Juno v0.5.5
  [b964fa9f] LaTeXStrings v1.0.3
  [50d2b5c4] Lazy v0.13.2
  [1914dd2f] MacroTools v0.4.5
  [e1d29d7a] Missings v0.4.0
  [9b87118b] PackageCompiler v0.6.3+ [`C:\Users\RTX2080\.julia\dev\PackageCompiler`]
  [1a8c2f83] Query v0.11.0
  [ce6b1742] RDatasets v0.6.1
  [2913bbd2] StatsBase v0.29.0
  [70df011a] TableReader v0.1.0 #master (https://github.com/bicycle1885/TableReader.jl.git)
  [e0df1984] TextParse v0.8.0
  [10745b16] Statistics

(v1.1) pkg> dev TextParse
  Updating registry at `C:\Users\RTX2080\.julia\registries\General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
   Cloning git-repo `https://github.com/JuliaComputing/TextParse.jl.git`
  Updating git-repo `https://github.com/JuliaComputing/TextParse.jl.git`
 Resolving package versions...
  Updating `C:\Users\RTX2080\.julia\environments\v1.1\Project.toml`
  [e0df1984] ↑ TextParse v0.8.0 ⇒ v0.8.0+ [`C:\Users\RTX2080\.julia\dev\TextParse`]
  Updating `C:\Users\RTX2080\.julia\environments\v1.1\Manifest.toml`
  [e0df1984] ↑ TextParse v0.8.0 ⇒ v0.8.0+ [`C:\Users\RTX2080\.julia\dev\TextParse`]

julia> using TextParse; @time a = csvread("c:/data/Performance_2016Q1.csv", '|', header_exists=false)
ERROR: MethodError: no method matching pointer(::SubString{TextParse.VectorBackedUTF8String}, ::Int64)
Closest candidates are:
  pointer(::String, ::Integer) at strings/string.jl:82
  pointer(::SubString{String}, ::Integer) at strings/substring.jl:105
  pointer(::TextParse.VectorBackedUTF8String, ::Integer) at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\VectorBackedStrings.jl:16
  ...
Stacktrace:
 [1] _substring at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\field.jl:397 [inlined]
 [2] tryparsenext(::TextParse.StringToken{String}, ::SubString{TextParse.VectorBackedUTF8String}, ::Int64, ::Int64, ::TextParse.LocalOpts{UInt8,UInt8,UInt8}) at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\field.jl:368
 [3] macro expansion at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\util.jl:27 [inlined]
 [4] tryparsenext(::Quoted{String,TextParse.StringToken{String},UInt8,UInt8}, ::SubString{TextParse.VectorBackedUTF8String}, ::Int64, ::Int64, ::TextParse.LocalOpts{UInt8,UInt8,UInt8}) at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\field.jl:493
 [5] macro expansion at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\util.jl:27 [inlined]
 [6] tryparsenext(::TextParse.Field{String,Quoted{String,TextParse.StringToken{String},UInt8,UInt8}}, ::SubString{TextParse.VectorBackedUTF8String}, ::Int64, ::Int64, ::TextParse.LocalOpts{UInt8,UInt8,UInt8}) at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\field.jl:682
 [7] macro expansion at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\util.jl:27 [inlined]
 [8] quotedsplit(::SubString{TextParse.VectorBackedUTF8String}, ::TextParse.LocalOpts{UInt8,UInt8,UInt8}, ::Bool, ::Int64, ::Int64) at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\csv.jl:671
 [9] quotedsplit(::SubString{TextParse.VectorBackedUTF8String}, ::TextParse.LocalOpts{UInt8,UInt8,UInt8}, ::Bool) at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\csv.jl:662
 [10] #_csvread_internal#26(::Bool, ::Char, ::Char, ::Nothing, ::Type, ::Type, ::Bool, ::Int64, ::Nothing, ::Nothing, ::Int64, ::Nothing, ::Bool, ::Array{String,1}, ::Array{String,1}, ::OrderedCollections.OrderedDict{Union{Int64, String},AbstractArray{T,1} where T}, ::Int64, ::Nothing, ::Array{Any,1}, ::String, ::Int64, ::typeof(TextParse._csvread_internal), ::TextParse.VectorBackedUTF8String, ::Char) at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\csv.jl:367
 [11] (::getfield(TextParse, Symbol("#kw##_csvread_internal")))(::NamedTuple{(:filename, :header_exists),Tuple{String,Bool}}, ::typeof(TextParse._csvread_internal), ::TextParse.VectorBackedUTF8String, ::Char) at .\none:0
 [12] (::getfield(TextParse, Symbol("##22#24")){Base.Iterators.Pairs{Symbol,Bool,Tuple{Symbol},NamedTuple{(:header_exists,),Tuple{Bool}}},String,Char})(::IOStream) at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\csv.jl:110
 [13] #open#310(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::getfield(TextParse, Symbol("##22#24")){Base.Iterators.Pairs{Symbol,Bool,Tuple{Symbol},NamedTuple{(:header_exists,),Tuple{Bool}}},String,Char}, ::String, ::Vararg{String,N} where N) at .\iostream.jl:369
 [14] open at .\iostream.jl:367 [inlined]
 [15] #_csvread_f#20 at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\csv.jl:107 [inlined]
 [16] #_csvread_f at .\none:0 [inlined]
 [17] #csvread#16(::Base.Iterators.Pairs{Symbol,Bool,Tuple{Symbol},NamedTuple{(:header_exists,),Tuple{Bool}}}, ::Function, ::String, ::Char) at C:\Users\RTX2080\.julia\packages\TextParse\IAMBB\src\csv.jl:78
 [18] (::getfield(TextParse, Symbol("#kw##csvread")))(::NamedTuple{(:header_exists,),Tuple{Bool}}, ::typeof(csvread), ::String, ::Char) at .\none:0
 [19] top-level scope at util.jl:156
```

### PR DESCRIPTION
**Pull request by davidanthoff**
Fixes #125.

This changes the guess logic so that string columns always end up as a quoted string. That is the goal, I'm not a 100% sure I made all the necessary changes for that.

I think we should always parse string columns as quoted. The reason is Excel: in a string column, it will only use quotes in rows where they are needed. So if there is a string column and in a row way down there is a `,` in the string, then Excel will _not_ use quotes up until that row, and then use quotes just for that one cell. Our current logic will detect such a column as string (without quote), and then cannot recover from the row later that does use quotes: it will come across a line like this `", bar"`, and now it will read the content of the first column as just `"`, which is incorrect.

Long story short, I think there are lots of CSV files out there that have string columns where some rows use quotes and others don't, and the only way to read those properly seems to just always treat string columns as potentially having quotes.

I still need to fix tests, but wanted to get @shashi's general reaction first before I put more time into this.

### CODE REVIEW

**Comment by davidanthoff** [comment]
@shashi tests now pass, would be great to hear what you think about this!

File: src/guesstype.jl [hunk 1]
```diff
@@ -108,17 +108,20 @@ function guesstoken(x, opts, @nospecialize(prev_guess)=Unknown(), nastrings=NA_S
                 return Numeric(promote_type(T, fieldtype(prev_guess)))
             else
                 # something like a date turned into a single number?
-                return StringToken(stringarraytype<:StringArray ? StrRange : String)
+                y1 = StringToken(stringarraytype<:StringArray ? StrRange : String)
+                return prevent_quote_wrap ? y1 : Quoted(y1, opts.quotechar, opts.escapechar)
```
Comment (shashi) [inline src/guesstype.jl]: what is the semantics of `prevent_quote_wrap` exactly? When is it useful?

Comment (shashi) [reply src/guesstype.jl]: Ah ok got it, it's in the recursive call.

**Comment by shashi** [comment]
looks good to me! :+1:
